src/html_dnd.ts

```diff
@@ -12,26 +11,62 @@ namespace dnd {
     const dragstartEvent = createEventWithDataTransfer("dragstart", dataTransfer);
     draggable.dispatchEvent(dragstartEvent);
 
-    // For the drop event. The list of items representing dragged data can be
-    // read, including the data. No new data can be added.
-    store.mode = "readonly";
-
-    const dragOverEvent = createEventWithDataTransfer("dragover", dataTransfer);
-    droppable.dispatchEvent(dragOverEvent);
-
-    const dropEvent = createEventWithDataTransfer("drop", dataTransfer);
-    droppable.dispatchEvent(dropEvent);
-
-    // For all other events. The formats and kinds in the drag data store list
-    // of items representing dragged data can be enumerated, but the data itself
-    // is unavailable and no new data can be added.
-    store.mode = "protected";
-
-    const dragendEvent = createEventWithDataTransfer("dragend", dataTransfer);
-    draggable.dispatchEvent(dragendEvent);
+    const doDragEnter = () => {
+      // Simulate dragenter
+      store.mode = "readonly";
+      const dragEnterEvent = createEventWithDataTransfer("dragenter", dataTransfer);
+      droppable.dispatchEvent(dragEnterEvent);
+    };
+
+    const doDragOver = () => {
+      // For the drop event. The list of items representing dragged data can be
+      // read, including the data. No new data can be added.
+
+      const dragOverEvent = createEventWithDataTransfer("dragover", dataTransfer);
+      droppable.dispatchEvent(dragOverEvent);
+    };
+
+    const doDrop = () => {
+      const dropEvent = createEventWithDataTransfer("drop", dataTransfer);
+      droppable.dispatchEvent(dropEvent);
+
+      // For all other events. The formats and kinds in the drag data store list
+      // of items representing dragged data can be enumerated, but the data itself
+      // is unavailable and no new data can be added.
+      store.mode = "protected";
+    };
+
+    const doDragEnd = () => {
+      const dragendEvent = createEventWithDataTransfer("dragend", dataTransfer);
+      draggable.dispatchEvent(dragendEvent);
+    };
+
+    if (dispatchTimeout) {
+      setTimeout(() => {
+        doDragEnter();
+
+        setTimeout(() => {
+          doDragOver();
+
+          setTimeout(() => {
+            doDrop();
+
+            setTimeout(() => {
+              doDragEnd();
+            }, dispatchTimeout);
+          }, dispatchTimeout);
+        }, dispatchTimeout);
+      }, dispatchTimeout);
+      // Sleep to allow for the 4 events to occur before expecting any results.
+      sleep(dispatchTimeout * 4);
```

Comment:
I don't want to use the sleep if it is possible. Because it make very high CPU usage.

I'm searching other ways that use "[Execute Async Script API](https://www.w3.org/TR/webdriver/#execute-async-script)" such as [`WebDriver#executeAsyncScript`](https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/index_exports_WebDriver.html#executeAsyncScript).

But I don't know that whether major WebDriver libraries support this API.
If major WebDriver libraries do not support them, I' gonna accept the Pull request.

Reply:
Fairly, we can use "Execute Async Script" API on major libraries:

| Library | Status for "Execute Async Script" API | Source |
|:--|:--|:--|
| selenium-webdriver | Supported | https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/index_exports_WebDriver.html#executeAsyncScript |
| WebdriverIO | Supported | http://webdriver.io/api/protocol/executeAsync.html |
| Nightwatch.js | Supported | http://nightwatchjs.org/api/executeAsync.html |
| Protractor | Supported | http://www.protractortest.org/#/api?view=webdriver.WebDriver.prototype.executeAsyncScript |